src/main/java/mops/businesslogic/DirectoryServiceImpl.java

```diff
@@ -0,0 +1,253 @@
+     * Creates the group root directory.
+     *
+     * @param account user credentials
+     * @param groupId the group id
+     * @return the directory created
+     */
+    @Override
+    public Directory createRootFolder(Account account, Long groupId) throws MopsException {
+
+        roleService.checkIfRole(account, groupId, ADMINISTRATOR);
+        Set<String> roleNames = permissionService.fetchRolesInGroup(groupId);
+        Set<DirectoryPermissionEntry> permissions = createDefaultPermissions(roleNames);
+        DirectoryPermissions permission = new DirectoryPermissions(permissions);
+        DirectoryPermissions rootPermissions = directoryPermissionsRepo.save(permission);
+        Directory directory = Directory.of(groupId.toString(), null, groupId, rootPermissions);
+        return directoryRepository.save(directory);
+    }
+
+    /**
+     * Creates a new folder inside a folder.
+     *
+     * @param account     user credentials
+     * @param parentDirId id of the parent folder
+     * @param dirName     name of the new folder
+     * @return id of the new folder
+     */
+    @Override
+    public Directory createFolder(Account account, long parentDirId, String dirName) throws MopsException {
+        Directory rootDirectory = fetchDirectory(parentDirId);
+        roleService.checkWritePermission(account, rootDirectory);
+        Directory directory = rootDirectory.createSubDirectory(dirName); //NOPMD// this is no violation of demeter's law
+        return directoryRepository.save(directory);
+    }
+
+    /**
+     * Deletes a folder.
+     *
+     * @param account user credential
+     * @param dirId   id of the folder to be deleted
+     * @return the parent id of the deleted folder
+     */
+    @Override
+    @SuppressWarnings("PMD.LawOfDemeter") //these are not violations of demeter's law
+    public Directory deleteFolder(Account account, long dirId) throws MopsException {
+        Directory directory = fetchDirectory(dirId);
+        roleService.checkDeletePermission(account, directory);
+
+        List<FileInfo> files = fileInfoService.fetchAllFilesInDirectory(dirId);
+        List<Directory> subFolders = getSubFolders(account, dirId);
+
+        if (!files.isEmpty() || !subFolders.isEmpty()) {
+            String errorMessage = String.format("The directory %s is not empty.", directory.getName());
+            throw new DeleteAccessPermissionException(errorMessage);
+        }
+
+        Directory parentDirectory = fetchDirectory(directory.getParentId());
+
+        directoryRepository.delete(directory);
+
+        return parentDirectory;
+    }
+
+    /**
+     * Searches a folder for files.
+     *
+     * @param account user credentials
+     * @param dirId   id of the folder to be searched
+     * @param query   wrapper object of the query parameter
+     * @return list of files
+     */
+    @Override
+    public List<FileInfo> searchFolder(Account account, long dirId, FileQuery query) {
+        return null;
+    }
```

Comment:
```suggestion
    public List<FileInfo> searchFolder(Account account, long dirId, FileQuery query) {
        return null; //TODO: Implement folder search
    }
```

Implementation missing - added TODO